Add stub for sort API and configurable built-in sorting options.

IMarvinTPA opened the pull request:
I'm not sure how to finish wiring up the API stub to point to the public add method in IngredientListElementComparator.
I also wish I could replicate the sort configuration panel in NEI instead of comma delimited string in the config that it has now.

This is a partial solution for https://github.com/mezz/JustEnoughItems/issues/549 that eliminates any external dependencies unlike my last try.

It is substantially similar, and I am open to suggestions on getting it done differently if needed.

Thank You,
IMarv

IMarvinTPA committed:
Add stub for sort API and configurable built-in sorting options.

IMarvinTPA commented:
Unstale bump comment.

mezz commented:
This looks pretty good but the code formatting style is very different from the rest of the project.
Please see how I name things with fairly long names, the project only uses tabs, and see how if/else are formatted, etc.

"invtweaks" should not be a default sorting method, if you need that hard-coded into JEI then I think the abstraction is not good.
Configuration options like this should be up to the player, I don't want want other mods to automatically change the sort order because the source of the change would be unknown and that is confusing to players.

IMarvinTPA commented:
I'll get the formatting and naming figured out.  (I'm horrible about white-space as long as it looks right for the tabs/spaces.)
I think the string default value doesn't get used, I've noticed that new instances don't match that string anyway.  The real default seems to be generated based on the order the options are created.

IMarvinTPA committed:
Modified for style, API tools and probable API timeline of events.

IMarvinTPA commented:
I pushed an update.  I still haven't really sussed out how to make a new API in your style.  I used Eclipse's ctrl-shift-f for the auto-formatter to fix the space/tab thing.  However, this seemed to move some stuff around for line length enforcement.  Let me know if this is undesirable.

IMarvinTPA commented:
I pushed an update.  I still haven't really sussed out how to make a new API in your style.  I used Eclipse's ctrl-shift-f for the auto-formatter to fix the space/tab thing.  However, this seemed to move some stuff around for line length enforcement.  Let me know if this is undesirable.

mezz commented on src/api/java/mezz/jei/api/IModPlugin.java:
```diff
@@ -1,16 +1,21 @@
 /**
- * The main class to implement to create a JEI plugin. Everything communicated between a mod and JEI is through this class.
- * IModPlugins must have the {@link JEIPlugin} annotation to get loaded by JEI.
+ * The main class to implement to create a JEI plugin. Everything communicated
```
Please do not reformat existing lines to the wrong format.
???

IMarvinTPA replied:
This is a side-effect of using ctrl-shift-F in eclipse to fix the spaces to tabs issue.  I did not anticipate this.  How do you ensure indenting is all tabs and not spaces?  I'll go and try to undo as many of these as I can find.

mezz commented on src/api/java/mezz/jei/api/IModPlugin.java:
```diff
@@ -45,11 +50,25 @@ default void register(IModRegistry registry) {
+	 *
+	 * @since JEI 4.?.0
+	 */
+	default void registerSorter(String name, Comparator<ItemStack> comparator) {
```
This class gets implemented by mods, so this method is in the wrong place.
This method should go in `IModRegistry`, which is implemented by JEI and passed to mod plugins.

IMarvinTPA replied:
Thank you.  I can do that.

IMarvinTPA committed:
Cleaned up autoformatting line wraps.
Fixed issues with comparator contracts.
Improved Tool detection.
Moved and implemented Sort API function.

IMarvinTPA commented:
This should do the trick.  Thank you!

mezz commented on src/api/java/mezz/jei/api/IModPlugin.java:
```diff
@@ -1,10 +1,13 @@
 
 /**
- * The main class to implement to create a JEI plugin. Everything communicated between a mod and JEI is through this class.
+ * The main class to implement to create a JEI plugin. Everything communicated between a mod and JEI is through this class. 
```
Please make sure lines like this are not changed, it complicates the git history when someone needs to look for the source of a change.

IMarvinTPA replied:
Ctrl-M is the devil.  I'll dos2unix the crap out of the files.

IMarvinTPA replied:
dos2unix and unix2dos were worse.  I found that comparing the documents with the originals would highlight the invisible character and I could delete it out of the comparison window and save.

mezz commented on src/main/java/mezz/jei/config/Config.java:
```diff
@@ -582,7 +627,8 @@ private static void updateBlacklist() {
 		boolean updated = false;
 
-		// deconstruct any mod-id blacklists into lower-level ones first. mod-id blacklist is deprecated
+		// deconstruct any mod-id blacklists into lower-level ones first. mod-id
```
should not be changed

mezz commented on src/main/java/mezz/jei/config/Config.java:
```diff
@@ -73,8 +74,8 @@ private Config() {
-		return values.overlayEnabled ||
-			KeyBindings.toggleOverlay.getKeyCode() == 0; // if there is no key binding to enable it, don't allow the overlay to be disabled
+		return values.overlayEnabled || 
+				KeyBindings.toggleOverlay.getKeyCode() == 0; // if there is no key binding to enable it, don't allow the overlay to be disabled
```
should not be changed

mezz commented on src/main/java/mezz/jei/config/Config.java:
```diff
@@ -662,16 +708,16 @@ private static void updateBlacklist() {
-	 *         The result is a positive integer if str1 is _numerically_ greater than str2.
-	 *         The result is zero if the strings are _numerically_ equal.
+	 * @return The result is a negative integer if str1 is _numerically_ less than str2. 
+	 *         The result is a positive integer if str1 is _numerically_ greater than str2. 
```
should not be changed

mezz commented on src/main/java/mezz/jei/config/ConfigValues.java:
```diff
@@ -25,4 +25,7 @@
 	public String filterText = "";
+
+	// Sort
+	public String itemSortlist = ""; // Gets populated dynamically.
```
Shouldn't this be a list of strings?

IMarvinTPA replied:
I've considered doing that.  Mostly, I haven't done it because editing the config with that style with the default GUI does not allow simply moving entries up and down, so you are forced to add blank lines, then copy/cut one option and paste it into the new spot.  I can try it if you really would like it, play around with the sample ones in forge's config area.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientFilter.java:
```diff
@@ -41,7 +42,7 @@
 	private final IngredientBlacklistInternal blacklist;
 	/**
-	 * indexed list of ingredients for use with the suffix trees
+	 * indexed list of ingredients for use with the suffix trees 
```
should not be changed

mezz commented on src/main/java/mezz/jei/ingredients/IngredientFilter.java:
```diff
@@ -171,9 +172,9 @@ public void updateHidden() {
-			ingredientHelper.isIngredientOnServer(ingredient) &&
-			(Config.isHideModeEnabled() || !Config.isIngredientOnConfigBlacklist(ingredient, ingredientHelper));
+		boolean visible = !blacklist.isIngredientBlacklistedByApi(ingredient, ingredientHelper) && 
+				ingredientHelper.isIngredientOnServer(ingredient) && 
```
should not be changed

mezz commented on src/main/java/mezz/jei/ingredients/IngredientFilter.java:
```diff
@@ -366,7 +371,7 @@ private IntSet getSearchResults(String token) {
 
 	/**
-	 * Efficiently get the elements contained in both sets.
+	 * Efficiently get the elements contained in both sets. 
```
should not be changed

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+import net.minecraft.item.ItemArmor;
+import net.minecraft.item.ItemHoe;
+import net.minecraft.item.ItemShears;
+import scala.Int;
```
importing scala can cause issues

IMarvinTPA replied:
I was only using it for Int.MaxValue().  Integer.MAX_VALUE should be used instead.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+import net.minecraft.item.ItemHoe;
+import net.minecraft.item.ItemShears;
+import scala.Int;
+import java.util.*;
```
do not use `*` imports

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementFactory.java:
```diff
@@ -23,7 +24,11 @@ private IngredientListElementFactory() {
+		try {
+			ingredientListElements.sort(IngredientListElementComparator.INSTANCE);
+		} catch (IllegalArgumentException ex) {
+			Log.get().error("Item sorting failed.  Aborting sort.", ex);
```
Why catch `IllegalArgumentException` in particular? Are there other errors that can happen?
We need some kind of fallback behavior when this happens so that players still get some basic sorting.

IMarvinTPA replied:
On a previous submission, I had been admonished for a generic catch.  This is the error thrown when timsort catches that a comparator violates the comparator contract.  I removed code that fell back to Quicksort.  Would you want that or force the sort to a basic default?  I favor a safe default.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+
+	}
+
+	private static <V> ItemStack getItemStack(IIngredientListElement<V> ingredientListElement) {
```
methods that can return null must be marked `@Nullable`

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
 public final class IngredientListElementComparator implements Comparator<IIngredientListElement> {
 	public static final IngredientListElementComparator INSTANCE = new IngredientListElementComparator();
 
+	public static ArrayList<SortEntry> entries = new ArrayList<SortEntry>();
```
1. Field declaration types should not be more specific than necessary.
2. We have java 8, use the diamond operator.

so this should be 
```java
public static List<SortEntry> entries = new ArrayList<>();
```

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
 
+	public static class SortEntry {
+		public String name;
+		// External interfaces only have access to ItemStacks, sorry fluids.
```
I think we should be able to sort any type of ingredient

IMarvinTPA replied:
Is IIngredientListElement is something accessible to modders with the API?
If so, there is no reason that the API can't offer that as well.

mezz replied:
The API can only access classes in the `mezz.jei.api` package, they can't access `IIngredientListElement`.

I think you could make the sorter take an `Object` instead of `ItemStack`, then they can check if it's an ingredient type they understand and sort it, or just return 0 if they don't want to deal with that type of ingredient.

IMarvinTPA replied:
What would I be sending the comparator?
After a little bit of review, I think I would be sending the IIngredientListElement.getIngredient() values.  I'm going to have to think.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
-			return 1;
-		} else {
-			return modName1.compareTo(modName2);
+		if (list.size() == 0) {
```
We should not make `IngredientListElementComparator` a singleton class. 
That way `list` can be passed into the constructor and we can be confident that anywhere we can access an instance of it, things are set correctly already.

IMarvinTPA replied:
Ok.

IMarvinTPA replied:
I made the empty constructor default to the config value.  There is also a constructor that takes a string and ignores the config.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+		}
+		ArrayList<SortEntry> templist = new ArrayList<SortEntry>(list);
+		templist.add(existingEntry);
+		list = templist;// concurrency
```
If we have concurrent modification at this step, some changes will be lost.
JEI is generally not done with parallel threads, so unless you are adding some parallel processing we should not be concerned with concurrent additions to the list.

IMarvinTPA replied:
These lines may be from the NEI code that I stole long ago.  I'll slim it down.

IMarvinTPA replied:
Converting the object away from the singleton mode removed these lines entirely.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+		return a == b ? 0 : a < b ? -1 : 1;
+	}
+	
+	public static int compareDouble(double a, double b) {
```
you can use `java.lang.Double#compare` for this

IMarvinTPA replied:
I cloned something I inherited.  I can do this.

mezz commented on src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java:
```diff
@@ -1,41 +1,483 @@
+		return null;
+	}
+
+	public static int compareInt(int a, int b) {
```
`java.lang.Integer#compare`

IMarvinTPA replied:
I inherited this line from the NEI code, I can clean this up.

mezz reviewed (changes requested):
Please take the time to look over every line difference in Github, it should be as perfect as you can make it. Immediately spotting many small errors lowers my confidence in the code and that makes the review long. I try to hold this codebase to a high standard because people depend on it to work in all cases, and modders will reference it for how to handle things the right way.

Once things look good, you will need to run some performance metrics to measure the impact of changing the sorting system, to ensure that the new code is still fast in very large mod packs. If you take a scientific approach to it then it should be convincing and I can merge with confidence. If you need help with designing a setup to take measurements let me know.

IMarvinTPA committed:
Cleaned up more autoformatting.
Polished some code.
Converted IngredientListElementComparator away from singleton pattern.
Added backup sorting in the event the custom version fails.

IMarvinTPA committed:
Cleaned up one more ctrl-M.

IMarvinTPA commented:
I've cleaned it up a bit.  I'm trying to figure out what to do for statistics.  There are no existing JEI commands, so a /JEI PerfTestSort command or some such isn't a viable solution.  I think there may be a demo version where I just let it do 1000 sorts on shuffled items while running StoneBlock or SevTech with various passes using different sorting options including one with the classic sort and recording total sort time (and not counting shuffle time.)

I feel like I should be putting more @Nullables in the comparator but I'm uncertain what that will do or how required it is.

I am uncertain as to how to handle the suggestion that the comparator accept "object".  Primarily, how do I know what to send the comparator if it isn't IIngredientListElement.  Further review seems to indicate IIngredientListElement.getIngredient().

I really appreciate your patience with me on this.

IMarvinTPA committed:
Added Dictionary Sort Option.  Just about to convert Comparator to Object from ItemStack.

IMarvinTPA committed:
The ItemStack comparator now has the fluids masquerade as buckets.  It also sends nulls to the comparator for better sorting behavior.
The comparator system now allows for an 'Object' comparator so there is an option for dealing with any type of Stack that needs sorted.
Added code for profiling the various comparators.

IMarvinTPA commented:
Note on the metrics.  Classic Sort is the old comparator.  "New Classic Sort" is with the new comparator set to a list that results in almost the same sort.  (Fluids are sorted with their buckets instead of the end.)
The inventory is randomized with Collections.Shuffle between sorts.  I believe this makes the lists worse than a default build since mods generally cluster their item types.  The "pre-sorted" lines do not have the shuffle step, and are basically the best-case scenario.

It seems that my new comparator with the new Default order is about 20 times slower than the old one.  But still less than a second.  The new comparator with the classic order is about 10 times slower.  The various other benchmarks are for each of the options in turn, with the DefaultOrderComparator appended.  These functions are only called when the game starts or the user changes the order.  The code for taking the metrics has been left in the factory class, but the line to call them has been commented out.

Metrics for SevTech:
[21:23:33] [main/INFO] [jei]: Building ingredient list...
[21:23:34] [main/INFO] [jei]: List of classes found: 
[21:23:34] [main/INFO] [jei]: net.minecraft.item.ItemStack, mekanism.api.gas.GasStack, net.minecraftforge.fluids.FluidStack
[21:23:38] [main/INFO] [jei]: Classic Sort benchmark took a total of 3.798 s for 100 executions.
[21:23:38] [main/INFO] [jei]: Classic Sort benchmark took an average of 37 ms.
[21:23:39] [main/INFO] [jei]: Classic Sort pre-sorted benchmark took a total of 677.8 ms for 100 executions.
[21:23:39] [main/INFO] [jei]: Classic Sort pre-sorted benchmark took an average of 6 ms.
[21:24:12] [main/INFO] [jei]: New Classic Sort benchmark took a total of 32.93 s for 100 executions.
[21:24:12] [main/INFO] [jei]: New Classic Sort benchmark took an average of 329 ms.
[21:24:16] [main/INFO] [jei]: New Classic Sort pre-sorted benchmark took a total of 3.606 s for 100 executions.
[21:24:16] [main/INFO] [jei]: New Classic Sort pre-sorted benchmark took an average of 36 ms.
[21:25:22] [main/INFO] [jei]: Default Sort benchmark took a total of 1.077 min for 100 executions.
[21:25:22] [main/INFO] [jei]: Default Sort benchmark took an average of 645 ms.
[21:25:27] [main/INFO] [jei]: Default Sort pre-sorted benchmark took a total of 4.931 s for 100 executions.
[21:25:27] [main/INFO] [jei]: Default Sort pre-sorted benchmark took an average of 49 ms.
[21:26:06] [main/INFO] [jei]: Tool Sort benchmark took a total of 38.72 s for 100 executions.
[21:26:06] [main/INFO] [jei]: Tool Sort benchmark took an average of 387 ms.
[21:26:10] [main/INFO] [jei]: Tool Sort pre-sorted benchmark took a total of 4.079 s for 100 executions.
[21:26:10] [main/INFO] [jei]: Tool Sort pre-sorted benchmark took an average of 40 ms.
[21:26:53] [main/INFO] [jei]: Melee Sort benchmark took a total of 41.91 s for 100 executions.
[21:26:53] [main/INFO] [jei]: Melee Sort benchmark took an average of 419 ms.
[21:26:58] [main/INFO] [jei]: Melee Sort pre-sorted benchmark took a total of 4.628 s for 100 executions.
[21:26:58] [main/INFO] [jei]: Melee Sort pre-sorted benchmark took an average of 46 ms.
[21:27:37] [main/INFO] [jei]: Armor Sort benchmark took a total of 38.74 s for 100 executions.
[21:27:37] [main/INFO] [jei]: Armor Sort benchmark took an average of 387 ms.
[21:27:41] [main/INFO] [jei]: Armor Sort pre-sorted benchmark took a total of 4.107 s for 100 executions.
[21:27:41] [main/INFO] [jei]: Armor Sort pre-sorted benchmark took an average of 41 ms.
[21:28:35] [main/INFO] [jei]: Dictionary Sort benchmark took a total of 52.80 s for 100 executions.
[21:28:35] [main/INFO] [jei]: Dictionary Sort benchmark took an average of 528 ms.
[21:28:39] [main/INFO] [jei]: Dictionary Sort pre-sorted benchmark took a total of 4.397 s for 100 executions.
[21:28:39] [main/INFO] [jei]: Dictionary Sort pre-sorted benchmark took an average of 43 ms.
[21:29:12] [main/INFO] [jei]: Minecraft Sort benchmark took a total of 32.69 s for 100 executions.
[21:29:12] [main/INFO] [jei]: Minecraft Sort benchmark took an average of 326 ms.
[21:29:16] [main/INFO] [jei]: Minecraft Sort pre-sorted benchmark took a total of 3.826 s for 100 executions.
[21:29:16] [main/INFO] [jei]: Minecraft Sort pre-sorted benchmark took an average of 38 ms.
[21:29:50] [main/INFO] [jei]: Mod Sort benchmark took a total of 33.13 s for 100 executions.
[21:29:50] [main/INFO] [jei]: Mod Sort benchmark took an average of 331 ms.
[21:29:54] [main/INFO] [jei]: Mod Sort pre-sorted benchmark took a total of 3.763 s for 100 executions.
[21:29:54] [main/INFO] [jei]: Mod Sort pre-sorted benchmark took an average of 37 ms.
[21:30:06] [main/INFO] [jei]: Name Sort benchmark took a total of 12.20 s for 100 executions.
[21:30:06] [main/INFO] [jei]: Name Sort benchmark took an average of 121 ms.
[21:30:09] [main/INFO] [jei]: Name Sort pre-sorted benchmark took a total of 2.458 s for 100 executions.
[21:30:09] [main/INFO] [jei]: Name Sort pre-sorted benchmark took an average of 24 ms.
[21:30:43] [main/INFO] [jei]: Damage Sort benchmark took a total of 34.15 s for 100 executions.
[21:30:43] [main/INFO] [jei]: Damage Sort benchmark took an average of 341 ms.
[21:30:47] [main/INFO] [jei]: Damage Sort pre-sorted benchmark took a total of 3.482 s for 100 executions.
[21:30:47] [main/INFO] [jei]: Damage Sort pre-sorted benchmark took an average of 34 ms.
[21:31:27] [main/INFO] [jei]: ID Sort benchmark took a total of 39.62 s for 100 executions.
[21:31:27] [main/INFO] [jei]: ID Sort benchmark took an average of 396 ms.
[21:31:31] [main/INFO] [jei]: ID Sort pre-sorted benchmark took a total of 4.306 s for 100 executions.
[21:31:31] [main/INFO] [jei]: ID Sort pre-sorted benchmark took an average of 43 ms.
[21:32:03] [main/INFO] [jei]: 'DefaultOrderComparator' Sort benchmark took a total of 31.48 s for 100 executions.
[21:32:03] [main/INFO] [jei]: 'DefaultOrderComparator' Sort benchmark took an average of 314 ms.
[21:32:07] [main/INFO] [jei]: 'DefaultOrderComparator' Sort pre-sorted benchmark took a total of 3.687 s for 100 executions.
[21:32:07] [main/INFO] [jei]: 'DefaultOrderComparator' Sort pre-sorted benchmark took an average of 36 ms.
[21:32:07] [main/INFO] [jei]: Building ingredient list took 8.565 min

Metrics from StoneBlock:
[21:43:02] [main/INFO] [jei]: Building ingredient list...
[21:43:02] [main/INFO] [jei]: List of classes found: 
[21:43:02] [main/INFO] [jei]: net.minecraft.item.ItemStack, crazypants.enderio.base.integration.jei.energy.EnergyIngredient, mekanism.api.gas.GasStack, net.minecraftforge.fluids.FluidStack
[21:43:07] [main/INFO] [jei]: Classic Sort benchmark took a total of 4.870 s for 100 executions.
[21:43:07] [main/INFO] [jei]: Classic Sort benchmark took an average of 48 ms.
[21:43:08] [main/INFO] [jei]: Classic Sort pre-sorted benchmark took a total of 623.0 ms for 100 executions.
[21:43:08] [main/INFO] [jei]: Classic Sort pre-sorted benchmark took an average of 6 ms.
[21:43:48] [main/INFO] [jei]: New Classic Sort benchmark took a total of 38.84 s for 100 executions.
[21:43:48] [main/INFO] [jei]: New Classic Sort benchmark took an average of 388 ms.
[21:43:53] [main/INFO] [jei]: New Classic Sort pre-sorted benchmark took a total of 5.097 s for 100 executions.
[21:43:53] [main/INFO] [jei]: New Classic Sort pre-sorted benchmark took an average of 50 ms.
[21:45:16] [main/INFO] [jei]: Default Sort benchmark took a total of 1.360 min for 100 executions.
[21:45:16] [main/INFO] [jei]: Default Sort benchmark took an average of 816 ms.
[21:45:23] [main/INFO] [jei]: Default Sort pre-sorted benchmark took a total of 6.363 s for 100 executions.
[21:45:23] [main/INFO] [jei]: Default Sort pre-sorted benchmark took an average of 63 ms.
[21:46:05] [main/INFO] [jei]: Tool Sort benchmark took a total of 41.72 s for 100 executions.
[21:46:05] [main/INFO] [jei]: Tool Sort benchmark took an average of 417 ms.
[21:46:10] [main/INFO] [jei]: Tool Sort pre-sorted benchmark took a total of 5.285 s for 100 executions.
[21:46:10] [main/INFO] [jei]: Tool Sort pre-sorted benchmark took an average of 52 ms.
[21:47:00] [main/INFO] [jei]: Melee Sort benchmark took a total of 48.46 s for 100 executions.
[21:47:00] [main/INFO] [jei]: Melee Sort benchmark took an average of 484 ms.
[21:47:05] [main/INFO] [jei]: Melee Sort pre-sorted benchmark took a total of 5.655 s for 100 executions.
[21:47:05] [main/INFO] [jei]: Melee Sort pre-sorted benchmark took an average of 56 ms.
[21:47:44] [main/INFO] [jei]: Armor Sort benchmark took a total of 38.36 s for 100 executions.
[21:47:44] [main/INFO] [jei]: Armor Sort benchmark took an average of 383 ms.
[21:47:49] [main/INFO] [jei]: Armor Sort pre-sorted benchmark took a total of 5.148 s for 100 executions.
[21:47:49] [main/INFO] [jei]: Armor Sort pre-sorted benchmark took an average of 51 ms.
[21:48:51] [main/INFO] [jei]: Dictionary Sort benchmark took a total of 1.016 min for 100 executions.
[21:48:51] [main/INFO] [jei]: Dictionary Sort benchmark took an average of 609 ms.
[21:48:57] [main/INFO] [jei]: Dictionary Sort pre-sorted benchmark took a total of 5.876 s for 100 executions.
[21:48:57] [main/INFO] [jei]: Dictionary Sort pre-sorted benchmark took an average of 58 ms.
[21:49:33] [main/INFO] [jei]: Minecraft Sort benchmark took a total of 34.68 s for 100 executions.
[21:49:33] [main/INFO] [jei]: Minecraft Sort benchmark took an average of 346 ms.
[21:49:38] [main/INFO] [jei]: Minecraft Sort pre-sorted benchmark took a total of 5.022 s for 100 executions.
[21:49:38] [main/INFO] [jei]: Minecraft Sort pre-sorted benchmark took an average of 50 ms.
[21:50:14] [main/INFO] [jei]: Mod Sort benchmark took a total of 35.28 s for 100 executions.
[21:50:14] [main/INFO] [jei]: Mod Sort benchmark took an average of 352 ms.
[21:50:19] [main/INFO] [jei]: Mod Sort pre-sorted benchmark took a total of 5.185 s for 100 executions.
[21:50:19] [main/INFO] [jei]: Mod Sort pre-sorted benchmark took an average of 51 ms.
[21:50:34] [main/INFO] [jei]: Name Sort benchmark took a total of 14.84 s for 100 executions.
[21:50:34] [main/INFO] [jei]: Name Sort benchmark took an average of 148 ms.
[21:50:37] [main/INFO] [jei]: Name Sort pre-sorted benchmark took a total of 2.831 s for 100 executions.
[21:50:37] [main/INFO] [jei]: Name Sort pre-sorted benchmark took an average of 28 ms.
[21:51:12] [main/INFO] [jei]: Damage Sort benchmark took a total of 34.31 s for 100 executions.
[21:51:12] [main/INFO] [jei]: Damage Sort benchmark took an average of 343 ms.
[21:51:16] [main/INFO] [jei]: Damage Sort pre-sorted benchmark took a total of 4.548 s for 100 executions.
[21:51:16] [main/INFO] [jei]: Damage Sort pre-sorted benchmark took an average of 45 ms.
[21:51:59] [main/INFO] [jei]: ID Sort benchmark took a total of 41.28 s for 100 executions.
[21:51:59] [main/INFO] [jei]: ID Sort benchmark took an average of 412 ms.
[21:52:04] [main/INFO] [jei]: ID Sort pre-sorted benchmark took a total of 5.817 s for 100 executions.
[21:52:04] [main/INFO] [jei]: ID Sort pre-sorted benchmark took an average of 58 ms.
[21:52:39] [main/INFO] [jei]: 'DefaultOrderComparator' Sort benchmark took a total of 34.08 s for 100 executions.
[21:52:39] [main/INFO] [jei]: 'DefaultOrderComparator' Sort benchmark took an average of 340 ms.
[21:52:44] [main/INFO] [jei]: 'DefaultOrderComparator' Sort pre-sorted benchmark took a total of 5.227 s for 100 executions.
[21:52:44] [main/INFO] [jei]: 'DefaultOrderComparator' Sort pre-sorted benchmark took an average of 52 ms.
[21:52:45] [main/INFO] [jei]: Building ingredient list took 9.716 min
masquerade

mezz commented:
I won’t have a chance to do a proper review until this weekend or so, but it seems like you’re on the right track!

IMarvinTPA committed:
Added significant debugging assistance for badly performing comparators.
Made the Item Panel Reload public to allow InvTweaks to trigger an update when the tree changes.

IMarvinTPA commented:
Take your time.  It gives me a chance to iron out any kinks I can find.

IMarvinTPA commented:
I am satisfied with where I am at this point.  Any additional changes will be at your direction.

IMarvinTPA committed:
Merge remote-tracking branch 'upstream/1.12' into 1.12

IMarvinTPA commented:
I have fixed the conflict from the bookmarks.  Those are neat.

mezz commented:
I'm working on cleaning this up so it can be merged.
There is still a lot to do so I may not finish right away.

IMarvinTPA commented:
I look forward to seeing what you do and learning from the differences.  The only real requests I have is to not change some of the function names/class names I referenced in InvTweaks so the existing versions will work with the JEI that can be used by it.

IMarvinTPA committed:
Merge branch '1.12' of https://github.com/mezz/JustEnoughItems into 1.12

IMarvinTPA committed:
Improved getSneakyItemStack to handle everything.

Also added fishing rods as a fake tool-class.

IMarvinTPA commented:
Hmm.  I was trying to update for making the getSneakyItemStack better.

IMarvinTPA committed:
getSneakyItemStack needs to be able to return nulls for empty stacks.

IMarvinTPA committed:
Remove Null ItemStacks from comparator processes.

IMarvinTPA commented:
Hi.  I hope you enjoyed the holidays.  Just curious on how you are doing with this.
Thanks!
IMarv

mezz commented:
Sorry I have very limited time lately.
A little while ago, I worked on updating this to match what I wanted, and I pushed the branch up today so you can see:
https://github.com/mezz/JustEnoughItems/commit/cf0600eef76798e938207613629a28649a32ef98
There is still a lot to do, but it shows the direction I want things to go.
This is almost certainly going to conflict with the changes for the 1.13-pre branch, so I am waiting to finish that first. It will be a while before this feature is ready.

IMarvinTPA commented:
Thank you for entertaining the change.  I look forward to it being a normal feature.

IMarvinTPA commented:
Not stale.  Just complicated.

IMarvinTPA commented:
Bump

IMarvinTPA commented:
Bump, updated for latest 1.12 branch.
